homework10/businessinsider/task_data_parsing/task_with_async.py

```diff
@@ -0,0 +1,168 @@
+        growth = float(
+            company_info.find_all("td")[9]
+            .find_all("span")[1]
+            .text.replace("%", "")
+            .replace("%", "")
+        )
+        company_name_href_growth.append({"name": name, "href": href, "growth": growth})
+    return company_name_href_growth
+
+
+async def get_company_full_info(company_name_href_growth, current_dollar_value) -> Dict:
+    base_url = "https://markets.businessinsider.com"
+    company_url = base_url + company_name_href_growth["href"]
+    soup = BeautifulSoup(await get_html_content(company_url), "lxml")
+    price_in_dollars = soup.find(class_="price-section__current-value").text.replace(
+        ",", ""
+    )
+    one_company_full_data = {
+        "name": company_name_href_growth["name"],
+        "code": soup.find(class_="price-section__category").find("span").text[2:],
+        "price": round(float(price_in_dollars) * current_dollar_value, 2),
+        "p_e_ratio": get_value_with_reg_exp(
+            "P/E Ratio", soup.find_all("div", class_="snapshot__data-item")
+        ),
+        "growth": company_name_href_growth["growth"],
+        "potential_profit": get_potential_profit(soup),
+    }
+    return one_company_full_data
+
+
+def ten_most_expensive(data):
+    sorted_data = sorted(data, key=lambda x: x["price"], reverse=True)[:10]
+    save_data_in_json(sorted_data, "largest", key="price")
+
+
+def ten_lowest_p_e_ratio(data):
+    sorted_data = sorted(
+        filter(lambda d: d["p_e_ratio"], data),
+        key=lambda x: x["p_e_ratio"],
+        reverse=False,
+    )[:10]
+    save_data_in_json(sorted_data, "smallest", key="p_e_ratio")
+
+
+def ten_largest_growth(data):
+    sorted_data = sorted(data, key=lambda x: x["growth"], reverse=True)[:10]
+    save_data_in_json(sorted_data, "largest", key="growth")
+
+
+def ten_largest_potential_profit(data):
+    sorted_data = sorted(
+        filter(lambda d: d["potential_profit"], data),
+        key=lambda x: x["potential_profit"],
+        reverse=True,
+    )[:10]
+    save_data_in_json(sorted_data, "largest", key="potential_profit")
+
+
+def save_data_in_json(sorted_data: List[Dict], description: str, key: str):
+    data_to_json_file = [
+        {"code": company["code"], "name": company["name"], f"{key}": company[key]}
+        for company in sorted_data
+    ]
+    json_file_name = "ten_companies_with_" + description + "_" + key + ".json"
+    with open(json_file_name, "w") as file:
+        file.write(json.dumps(data_to_json_file, indent=4))
+
+
+async def main():
+    url_start = "https://markets.businessinsider.com/index/components/s&p_500?p="
+    ten_pages = [url_start + str(value) for value in range(1, 11)]
+    tasks_1 = [
+        asyncio.create_task(fetch_company_name_href_and_growth(url))
+        for url in ten_pages
+    ]
+    data_from_all_pages = await asyncio.gather(*tasks_1)
+    current_dollar_value = await get_current_dollar_value()
```

Comment:
How cold you optimize this line?

Reply:
Maybe you mean smth like this?
`current_dollar_value = await asyncio.gather(get_current_dollar_value(session))`

Reply:
You can try to get ideas in that post:
https://teams.microsoft.com/l/message/19:b2ee08adccab42d6921c3f464ee414aa@thread.tacv2/1618436237138?tenantId=b41b72d0-4e9f-4c26-8a69-f949f367c91d&groupId=7a62e404-a944-4eac-a988-fc327fabd309&parentMessageId=1618436237138&teamName=EPAM%20Python%20Feb%2721&channelName=%D0%9E%D0%B1%D1%89%D0%B8%D0%B9&createdTime=1618436237138